leo_filters/src/odom_filter.cpp

```diff
@@ -0,0 +1,176 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include <chrono>
+#include <cmath>
+
+#include "rclcpp/rclcpp.hpp"
+
+#include "odom_filter.hpp"
+
+using namespace std::chrono_literals;
+using std::placeholders::_1;
+using std::placeholders::_2;
+
+namespace leo_filters
+{
+OdomFilter::OdomFilter(rclcpp::NodeOptions options)
+: Node("odom_filter", options),
+  param_listener_(get_node_parameters_interface())
+{
+  params_ = param_listener_.get_params();
+  tf_frame_prefix_ = declare_parameter("tf_frame_prefix", tf_frame_prefix_);
+
+  imu_sub_ = create_subscription<sensor_msgs::msg::Imu>(
+    "imu/data", rclcpp::QoS(5).best_effort(),
+    std::bind(&OdomFilter::imu_callback, this, _1));
+  wheel_odom_sub_ = create_subscription<nav_msgs::msg::Odometry>(
+    "wheel_odom_with_covariance", rclcpp::QoS(5).best_effort(),
+    std::bind(&OdomFilter::odom_callback, this, _1));
+
+  broadcaster_ = std::make_unique<tf2_ros::TransformBroadcaster>(*this);
+
+  client_cb_group_ =
+    create_callback_group(rclcpp::CallbackGroupType::MutuallyExclusive);
+  reset_odom_client_ = create_client<std_srvs::srv::Trigger>(
+    "firmware/reset_odometry", rclcpp::ServicesQoS(),
+    client_cb_group_);
+  reset_odom_srv_ = create_service<std_srvs::srv::Trigger>(
+    "reset_odometry",
+    std::bind(&OdomFilter::reset_odom_callback, this, _1, _2));
+
+  odom_merged_pub_ =
+    create_publisher<nav_msgs::msg::Odometry>("merged_odom", 10);
+
+  odom_merged_timer_ = create_wall_timer(
+    10ms, std::bind(&OdomFilter::odom_merged_callback, this));
+
+  RCLCPP_INFO(get_logger(), "Started node");
+}
+
+void OdomFilter::odom_merged_callback()
+{
+  if (!imu_received_ || !odom_received_) {
+    return;
+  }
+
+  check_dynamic_parameters();
+  odom_merged_msg_.child_frame_id =
+    tf_frame_prefix_ + params_.robot_frame_id;
+  odom_merged_msg_.header.stamp = get_clock()->now();
+
+  double vel_x = odom_merged_msg_.twist.twist.linear.x;
+  double vel_y = odom_merged_msg_.twist.twist.linear.y;
+
+  const double move_x =
+    vel_x * std::cos(odom_merged_yaw_) - vel_y * std::sin(odom_merged_yaw_);
+  const double move_y =
+    vel_x * std::sin(odom_merged_yaw_) + vel_y * std::cos(odom_merged_yaw_);
+
+  odom_merged_msg_.pose.pose.position.x += move_x * 0.01;
+  odom_merged_msg_.pose.pose.position.y += move_y * 0.01;
+
+  odom_merged_yaw_ += odom_merged_msg_.twist.twist.angular.z * 0.01;
+
+  if (odom_merged_yaw_ > 2.0 * PI) {
+    odom_merged_yaw_ -= 2.0 * PI;
+  } else if (odom_merged_yaw_ < 0.0) {
+    odom_merged_yaw_ += 2.0 * PI;
+  }
+  odom_merged_msg_.pose.pose.orientation.z = std::sin(odom_merged_yaw_ * 0.5);
+  odom_merged_msg_.pose.pose.orientation.w = std::cos(odom_merged_yaw_ * 0.5);
+
+  odom_merged_pub_->publish(odom_merged_msg_);
+
+  if (params_.publish_tf) {
+    tf_msg_.header.stamp = get_clock()->now();
+    tf_msg_.header.frame_id = odom_merged_msg_.header.frame_id;
```

Comment:
Use the same stamp for tf messages (just copy the whole header)
```suggestion
    tf_msg_.header = odom_merged_msg_.header;
```